In [2]:
# The first step of using numpy is to tell python to import it
import numpy as np
import pandas as pd

import says to import specific package, then name of package, then as something, so np in this case
what this does is allow for the package as np, this allows for np to be a shorthand

---


to use packages they need to be installed first


### 2.1 NumPy Arrays

**NumPy Array**
* An array is a data structure that stores values of same data type.
* While python lists can contain values corresponding to different data types, arrays in python can only contain values corresponding to the same data type.
* However python lists fail to deliver the performance required while computing large sets of numerical data. To address this issue we use NumPy arrays.
* We can create NumPy arrays by converting a list to an array.


In [ ]:
# defining a list of different car companies or string elements. can only be one data type, in this case as string
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [0, -1, -2, 7, 3]

In [ ]:
# connverting the list arr_str to a NumPy array
np_arr_str = np.array(arr_str)


# connverting the list arr_num to a NumPy array
np_arr_num = np.array(arr_num)

# checking the output
print('Numpy Array (arr_str): ',np_arr_str)
print('Numpy Array (arr_num): ',np_arr_num)

Numpy Array (arr_str):  ['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
Numpy Array (arr_num):  [ 0 -1 -2  7  3]


In [ ]:
arr_num2 = [0, -1, -2, 7, 3]
a = np.array(arr_num)

The resuts look similar to a list but arr_str and arr_num have been converted to NumPy arrays. Let's check the data type to confirm this.

In [ ]:
# printing the data type of lists
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


* The above output confirms that both the lists were successfully converted to arrays

**NumPy Matrix**

* A matrix is a two-dimensional data structure where elements are arranged into rows and columns.
* A matrix can be created by using list of lists

In [ ]:
# let's say we have information of different number of cylinders in a car and we want to display them in a matrix format. lists of lists
matrix = np.array([[1,2,1],[4,5,9],[1,8,9]])
print(matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [ ]:
print('Data type of matrix: ',type(matrix))

Data type of matrix:  <class 'numpy.ndarray'>


* We see that all the NumPy objects have data type as ndarray

### 2.2 NumPy Functions

**There are different ways to create NumPy arrays using the functions available in NumPy library**

**Using np.arange() function**
* The np.arange() function returns an array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. start is included but stop is excluded.
* It has the following paramaters:
  * start : start of interval range. By default start = 0
  * stop  : end of interval range
  * step  : step size of interval. By default step size = 1

In [ ]:
arr2  = np.arange(start = 0, stop = 10) # 10 will be excluded from the output
print(arr2)

# or this works as well too

arr2  = np.arange(0,10)
print(arr2)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 9]


In [ ]:
# adding a step size of 5 to create an array
arr3  = np.arange(start = 0, stop = 20, step = 5)
print(arr3)
# or
arr3  = np.arange(0, 20, 5)
print(arr3)

[ 0  5 10 15]
[ 0  5 10 15]


In [ ]:
vec = np.arange(10, 20, 3)
print(vec)

[10 13 16 19]


**Using np.linspace() function**
* The np.linspace() function returns numbers which are evenly distributed with respect to interval. Here the start and stop both are included.            
*It has the following parameters:              
 * start: start of interval range. By default start = 0
 * stop: end of interval range
 * num : No. of samples to generate. By default num = 50

In [ ]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
# one of the few exceptions that stops where it tells it to stop
matrix2
# this is actually all just one row but it looks 2 dimensional just to make reading easier
# what this table does is put 50 even spaced numbers between 0 and 5 with 0 and 5 included

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

In [ ]:
vec = np.linspace(10,20,3)
print(vec)


[10. 15. 20.]


**How are these values getting generated?**

The step size or the difference between each element will be decided by the following formula:

**(stop - start) / (total elements - 1)**

So, in this case:
(5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be 0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082 +0.10204082, and so on.

In [ ]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,10)
matrix3

array([10.        , 11.11111111, 12.22222222, 13.33333333, 14.44444444,
       15.55555556, 16.66666667, 17.77777778, 18.88888889, 20.        ])

In [ ]:
# generating 11 evenly spaced values between 10 and 20 gives the right number wth one between each
matrix3 = np.linspace(10,20,11)
matrix3

array([10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20.])

In [ ]:
vec1 = np.linspace(20,50,4)
vec2 = np.linspace(100,120,3)
print(vec1 + vec2)

ValueError: operands could not be broadcast together with shapes (4,) (3,) 

In [ ]:
vec = np.arange(2,5,1)
print(1/vec)
print(vec**2)

[0.5        0.33333333 0.25      ]
[ 4  9 16]


In [ ]:
mat1 = np.arange(1,5).reshape(2,2)
mat2 = np.eye(2)
print(mat1/mat2)

[[ 1. inf]
 [inf  4.]]


<ipython-input-17-0e6adcba2ca7>:3: RuntimeWarning: divide by zero encountered in divide
  print(mat1/mat2)


**Similarly we can create matrices using the functions available in NumPy library**

**Using np.zeros()**

* The np.zeros() is a function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix filled with zeros of the given shape.
* It has the following parameters:    
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [ ]:
matrix4 = np.zeros([3,5])
# this is a matrix completly made of 0s but x rows adn y columsn so 3x5 in this case
matrix4

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

**Using np.ones()**

* The np.ones() is another function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix of given shape and type, filled with ones.
* It has the following parameters:  
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [ ]:
matrix5 = np.ones([3,5])
# same as above with ones
matrix5

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

**Using np.eye()**
* The np.eye() is a function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix with ones on the diagonal and zeros elsewhere.
* It has the following parameters:
  * n: Number of rows and columns in the output matrix
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [ ]:
matrix6 = np.eye(5)
# so shorthand for identical and used eye
# numbers and rows and columns are the same. 1s are from top left to bottom right corner (0,0) points to (x,x) point
matrix6

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

**We can also convert a one dimension array to a matrix. This can be done by using the np.reshape() function.**

* The shape of an array basically tells the number of elements and dimensions of the array. Reshaping a Numpy array simply means changing the shape of the given array.
* By reshaping an array we can add or remove dimensions or change number of elements in each dimension.
* In order to reshape a NumPy array, we use the reshape method with the given array.
* **Syntax:** array.reshape(shape)
  * shape: a tuple given as input, the values in tuple will be the new shape of the array.

In [ ]:
# defining an array with values 0 to 9
arr4 = np.arange(0,10)
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
# reshaping the array arr4 to a 2 x 5 matrix
# a function inside an object is called a method
arr4_reshaped = arr4.reshape((2,5))
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [ ]:
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
# reshaping the array arr4 to a 2 x 6 matrix
arr4.reshape((2,6))

ValueError: cannot reshape array of size 10 into shape (2,6)

* This did not work because we have 10 elements which we are trying to fit in a 2 X 6 shape which will require 12 elements.

**NumPy can also perform a large number of different mathematical operations and it provides different functions to do so.**

NumPy provides:
1. Trigonometric functions
2. Exponents and Logarithmic functions
3. Functions for arithmetic operations between arrays and matrices

**Trigonometric functions**

In [ ]:
print('Sine Function:',np.sin(4))
print('Cosine Function:',np.cos(4))
print('Tan Function',np.tan(4))

Sine Function: -0.7568024953079282
Cosine Function: -0.6536436208636119
Tan Function 1.1578212823495775


**Exponents and Logarithmic functions**

* Exponents

In [ ]:
np.exp(2)

7.38905609893065

In [ ]:
arr5 = np.array([2,4,6])
np.exp(arr5)

array([  7.3890561 ,  54.59815003, 403.42879349])

* Logarithms

In [ ]:
# by default NumPy takes the base of log as e
np.log(2)

0.6931471805599453

In [ ]:
np.log(arr5)

array([0.69314718, 1.38629436, 1.79175947])

In [ ]:
## log with base 10
np.log10(8)

0.9030899869919435

**Arithmetic Operations on arrays**

In [ ]:
# arithmetic on lists

l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)
# this does not behave as you would expect!


[1, 2, 3, 4, 5, 6]


In [ ]:
# we can +-*/ arrays together

# defining two arrays
arr7 = np.arange(1,6)
print('arr7:', arr7)

arr8 = np.arange(3,8)
print('arr8:', arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [ ]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **, NOT ^!!! ^ does something completely different!


Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


**Operations on Matrices**

In [ ]:
matrix7 = np.arange(1,10).reshape(3,3)
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
print('Addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: divide by zero encountered in true_divide
  after removing the cwd from sys.path.


* RuntimeWarning: Errors which occur during program execution(run-time) after successful compilation are called run-time errors.
* One of the most common run-time error is division by zero also known as Division error.
* Due to division by zero error, we are getting inf (infinity) values because 1/0 is not a defined operation.

**Linear algebra matrix multiplication**

In [ ]:
matrix9 = np.arange(1,10).reshape(3,3)
print('First Matrix: \n',matrix9)

matrix10 = np.arange(11,20).reshape(3,3)
print('Second Matrix: \n',matrix10)
print('')
# taking linear algebra matrix multiplication (some may have heard this called the dot product)
print('Multiplication: \n', matrix9 @ matrix10)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


**Transpose of a matrix**

In [ ]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [ ]:
# taking transpose of matrix
np.transpose(matrix9)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [ ]:
# another way of taking a transpose
matrix9.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

**Function to find minimum and maximum values**

In [ ]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [ ]:
print('Minimum value: ',np.min(matrix9))

Minimum value:  1


In [ ]:
print('Maximum value: ',np.max(matrix9))

Maximum value:  9


**Function to generate random samples**

**Using np.random.rand function**

* The np.random.rand returns a random NumPy array whose element(s) are drawn randomly from the uniform distribution over [0,1). (including 0 but excluding 1).
* **Syntax** - np.random.rand(d0,d1)
  * d0,d1 – It represents the dimension of the required array given as int, where d1 is optional.

In [ ]:
# Generating random values in an array
rand_mat = np.random.rand(5)
print(rand_mat)

[0.77228845 0.37222006 0.32151418 0.11785626 0.52418771]


In [ ]:
# * Generating random values in a matrix
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)

[[0.48192504 0.49143874 0.04261904 0.16733493 0.21889742]
 [0.53714054 0.04059588 0.82157237 0.28154908 0.14718889]
 [0.23271254 0.06610744 0.82847405 0.931699   0.51780884]
 [0.80194863 0.4714679  0.36169667 0.71215477 0.53264981]
 [0.50439349 0.71730351 0.36758533 0.38440781 0.38511696]]


**Using np.random.randn function**

* The np.random.randn returns a random numpy array whose sample(s) are drawn randomly from the standard normal distribution (Mean as 0 and standard deviation as 1)

* **Syntax** - np.random.randn(d0,d1)
  * d0,d1 – It represents the dimension of the output, where d1 is optional.

In [ ]:
# Generating random values in an array
rand_mat2 = np.random.randn(5)
print(rand_mat2)

[ 1.28449866 -1.13347145 -0.7653113   1.572269   -0.24304349]


In [ ]:
# Generating random values in a matrix
rand_mat2 = np.random.randn(5,5)
print(rand_mat2)

[[ 0.20118246 -0.03672446 -0.052403   -0.47611353  1.10730079]
 [ 0.48381972 -0.19107893  1.05540306  0.47520419 -0.47128826]
 [-2.01034073  2.30599323 -0.02722149 -0.13580281  0.65920198]
 [-0.36713617  0.51046851 -0.7530001  -0.79756846  0.70224427]
 [ 0.17033068  0.55570677  1.68680188 -0.35406448  0.23750957]]


In [ ]:
# Let's check the mean and standard deviation of rand_mat2
print('Mean:',np.mean(rand_mat2))
print('Standard Deviation:',np.std(rand_mat2))

Mean: 0.11722991137038243
Standard Deviation: 0.883066783507808


*  We observe that the mean is very close to 0 and standard deviation is very close to 1.

**Using np.random.randint function**

* The np.random.randint returns a random numpy array whose element(s) are drawn randomly from low (inclusive) to the high (exclusive) range.

* **Syntax** - np.random.randint(low, high, size)

  * low – It represents the lowest inclusive bound of the distribution from where the sample can be drawn.
  * high – It represents the upper exclusive bound of the distribution from where the sample can be drawn.
  * size – It represents the shape of the output.

In [ ]:
# Generating random values in an array
rand_mat3 = np.random.randint(1,5,10)
print(rand_mat3)

[3 2 3 1 2 4 3 3 4 4]


In [ ]:
# Generating random values in a matrix
rand_mat3 = np.random.randint(1,10,[5,5])
print(rand_mat3)

[[1 4 5 5 8]
 [9 1 7 4 1]
 [9 2 7 1 8]
 [7 7 6 3 9]
 [9 9 7 4 7]]


### 2.3 Accessing the entries of a Numpy Array

In [ ]:
# let's generate an array with 10 random values
rand_arr = np.random.randn(10)
print(rand_arr)

[-0.61610084  0.256255   -1.02474372  1.69663775  0.14830608  0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


* Accessing one element from an array

In [ ]:
# accessing the 6 th entry of rand_arr
print(rand_arr[6])

-0.5845112936165019


* Accessing multiple elements from an array

In [ ]:
# we can access multiple entries at once using
print(rand_arr[4:9])

[ 0.14830608  0.0587465  -0.58451129  0.09827507  1.38926594]


In [ ]:
# we can also access multiple non-consecutive entries using np.arange
print('Index of values to access: ',np.arange(3,10,3))
print(rand_arr[np.arange(3,10,3)])

Index of values to access:  [3 6 9]
[ 1.69663775 -0.58451129  0.52114238]


**Accessing arrays using logical operations**

In [ ]:
print(rand_arr)

[-0.61610084  0.256255   -1.02474372  1.69663775  0.14830608  0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
rand_arr>0

array([False,  True, False,  True,  True,  True, False,  True,  True,
        True])

In [ ]:
# accessing all the values of rand_arr which are greater than 0
print('Values greater than 0: ',rand_arr[rand_arr>0])

# accessing all the values of rand_arr which are less than 0
print('Values less than 0: ',rand_arr[rand_arr<0])

Values greater than 0:  [0.256255   1.69663775 0.14830608 0.0587465  0.09827507 1.38926594
 0.52114238]
Values less than 0:  [-0.61610084 -1.02474372 -0.58451129]


**Accessing the entries of a Matrix**

In [ ]:
# let's generate an array with 10 random values
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[ 0.66008467 -0.91860835 -0.91768744 -0.56565779 -0.19013084]
 [ 0.68692112  1.20042327 -0.55648032  0.51675983 -1.13325252]
 [ 1.76883638 -0.85498119  1.89998922 -1.0658905   1.50531893]
 [ 0.77388464 -1.24206763  0.75631041  1.01940576  0.89747448]
 [ 0.37852758 -1.36846366 -0.91566544 -0.47943544 -0.20226054]]


In [ ]:
# acessing the second row of the rand_mat
rand_mat[1]

array([ 0.68692112,  1.20042327, -0.55648032,  0.51675983, -1.13325252])

In [ ]:
# acessing third element of the second row
print(rand_mat[1][2])

#or

print(rand_mat[1,2])

-0.5564803167132045
-0.5564803167132045


In [ ]:
# accessing first two rows with second and third column
print(rand_mat[0:2,1:3])

[[-0.91860835 -0.91768744]
 [ 1.20042327 -0.55648032]]


**Accessing matrices using logical operations**

In [ ]:
print(rand_mat)

[[ 0.66008467 -0.91860835 -0.91768744 -0.56565779 -0.19013084]
 [ 0.68692112  1.20042327 -0.55648032  0.51675983 -1.13325252]
 [ 1.76883638 -0.85498119  1.89998922 -1.0658905   1.50531893]
 [ 0.77388464 -1.24206763  0.75631041  1.01940576  0.89747448]
 [ 0.37852758 -1.36846366 -0.91566544 -0.47943544 -0.20226054]]


In [ ]:
# accessing all the values of rand_mat which are greater than 0
print('Values greater than 0: \n ',rand_mat[rand_mat>0])

# accessing all the values of rand_mat which are less than 0
print('Values less than 0: \n',rand_mat[rand_mat<0])

Values greater than 0: 
  [0.66008467 0.68692112 1.20042327 0.51675983 1.76883638 1.89998922
 1.50531893 0.77388464 0.75631041 1.01940576 0.89747448 0.37852758]
Values less than 0: 
 [-0.91860835 -0.91768744 -0.56565779 -0.19013084 -0.55648032 -1.13325252
 -0.85498119 -1.0658905  -1.24206763 -1.36846366 -0.91566544 -0.47943544
 -0.20226054]


**Modifying the entries of an Array**

In [ ]:
print(rand_arr)


[-0.61610084  0.256255   -1.02474372  1.69663775  0.14830608  0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
# let's change some values in an array!
# changing the values of index value 3 and index value 4 to 5
rand_arr[3:5] = 5
print(rand_arr)

[-0.61610084  0.256255   -1.02474372  5.          5.          0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
# changing the values of index value 0 and index value 1 to 2 and 3 respectively
rand_arr[0:2] = [2,3]
print(rand_arr)

[ 2.          3.         -1.02474372  5.          5.          0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
# modify entries using logical references
rand_arr[rand_arr>0] = 65
rand_arr

array([65.        , 65.        , -1.02474372, 65.        , 65.        ,
       65.        , -0.58451129, 65.        , 65.        , 65.        ])

**Modifying the entries of a Matrix**

In [ ]:
print(rand_mat3)

[[1 5 2 1 2]
 [6 1 7 1 8]
 [6 6 5 2 5]
 [4 3 7 3 1]
 [2 6 1 1 5]]


In [ ]:
# changing the values of the 4th and 5th element of the second and third rows of the matrix to 0
print('Matrix before modification: \n',rand_mat3)
rand_mat3[1:3,3:5] = 0
print('Matrix after modification: \n',rand_mat3)

Matrix before modification: 
 [[1 5 2 1 2]
 [6 1 7 1 8]
 [6 6 5 2 5]
 [4 3 7 3 1]
 [2 6 1 1 5]]
Matrix after modification: 
 [[1 5 2 1 2]
 [6 1 7 0 0]
 [6 6 5 0 0]
 [4 3 7 3 1]
 [2 6 1 1 5]]


In [ ]:
# extracting the first 2 rows and first 3 columns from the matrix
sub_mat = rand_mat[0:2,0:3]
print(sub_mat)

[[ 0.66008467 -0.91860835 -0.91768744]
 [ 0.68692112  1.20042327 -0.55648032]]


In [ ]:
# changing all the values of the extracted matrix to 3
sub_mat[:] = 3
print(sub_mat)

[[3. 3. 3.]
 [3. 3. 3.]]


In [ ]:
# what happened to rand_mat when we change sub_mat?
rand_mat

array([[ 3.        ,  3.        ,  3.        , -0.56565779, -0.19013084],
       [ 3.        ,  3.        ,  3.        ,  0.51675983, -1.13325252],
       [ 1.76883638, -0.85498119,  1.89998922, -1.0658905 ,  1.50531893],
       [ 0.77388464, -1.24206763,  0.75631041,  1.01940576,  0.89747448],
       [ 0.37852758, -1.36846366, -0.91566544, -0.47943544, -0.20226054]])

In [ ]:
# to prevent this behavior we need to use the .copy() method when we assign sub_mat
# this behavior is the source of MANY errors for early python users!!!

rand_mat = np.random.randn(5,5)
print(rand_mat)
sub_mat = rand_mat[0:2,0:3].copy()
sub_mat[:] = 3
print(sub_mat)
print(rand_mat)

[[ 2.0112384   0.49785818 -1.25183808  1.68261636  0.51405209]
 [ 0.19861572 -1.02405923  1.47588997  0.9095827   1.44913937]
 [ 1.17797906  2.2230628   1.0897073  -0.61658736  1.30247375]
 [-0.72899737  0.35296055 -1.28267073 -0.12390565  0.74816716]
 [ 1.54172729  0.23114564 -0.06203034 -0.00718632 -0.32182532]]
[[3. 3. 3.]
 [3. 3. 3.]]
[[ 2.0112384   0.49785818 -1.25183808  1.68261636  0.51405209]
 [ 0.19861572 -1.02405923  1.47588997  0.9095827   1.44913937]
 [ 1.17797906  2.2230628   1.0897073  -0.61658736  1.30247375]
 [-0.72899737  0.35296055 -1.28267073 -0.12390565  0.74816716]
 [ 1.54172729  0.23114564 -0.06203034 -0.00718632 -0.32182532]]


### 2.4 Saving and Loading a NumPy array

**Let's save some NumPy objects on the disk for use later!**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# creating a random matrices
randint_matrix1 = np.random.randint(1,10,10).reshape(2,5)
print(randint_matrix1)
print('')
randint_matrix2 = np.random.randint(10,20,10).reshape(2,5)
print(randint_matrix2)

[[7 2 9 7 2]
 [6 6 5 7 2]]

[[10 16 12 17 18]
 [19 12 11 13 17]]


**Using np.save() function**

In [ ]:
np.save('/content/drive/MyDrive/Python Course/saved_file_name',randint_matrix1)

**Using np.savez() function**

In [ ]:
np.savez('/content/drive/MyDrive/Python Course/multiple_files',randint_matrix1=randint_matrix1,randint_matrix2=randint_matrix2)

* The files will be saved in the directory where the Jupyter Notebook is located.
* With np.save() function, we can save an array/matrix to a NumPy .npy format.
* np.savez() function has an advantage over np.save() function because with np.savez(), we can store several arrays/matrices into a single file in uncompressed .npz format.

In [ ]:
# now let's load it
loaded_arr = np.load('/content/drive/MyDrive/Python Course/saved_file_name.npy')
loaded_multi = np.load('/content/drive/MyDrive/Python Course/multiple_files.npz')

print(loaded_arr)
print('')
print(loaded_multi)

* We see that .npy file has been loaded but the .npz file is returning a memory location.
* Let's see how to load the values stored in .npz file.

In [ ]:
print('1st Matrix: \n',loaded_multi['randint_matrix1'])
print('2nd Matrix: \n',loaded_multi['randint_matrix2'])

new_matrix  = loaded_multi['randint_matrix1']
print('New Matrix: \n',new_matrix)

In [ ]:
# we can also save/load text files...but only single variables
np.savetxt('/content/drive/MyDrive/Python Course/text_file_name.txt',randint_matrix1,delimiter=',')
rand_mat_txt = np.loadtxt('/content/drive/MyDrive/Python Course/text_file_name.txt',delimiter=',')
print(randint_matrix1)
print('')
print(rand_mat_txt)

In [ ]:
np.array([1,10,21,35,72])

array([ 1, 10, 21, 35, 72])

In [ ]:
np.array([[9, 8, 7], [6, 5, 4], [3, 2, 1]])

array([[9, 8, 7],
       [6, 5, 4],
       [3, 2, 1]])

In [ ]:
np.array([[9, 8, 7, 6, 5, 4, 3, 2, 1]])

array([[9, 8, 7, 6, 5, 4, 3, 2, 1]])

In [ ]:
np.arange(start=1, stop=10, step=1).reshape(3,3)

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

array_1 = np.linspace(0,20,6)
array_2 = np.arange(0,20,6)
print('array_1')
print('array_2')

In [ ]:
array_1 = np.linspace(0,20,6)
array_2 = np.arange(0,20,6)
print(array_1)
print(array_2)

[ 0.  4.  8. 12. 16. 20.]
[ 0  6 12 18]


In [ ]:
mat = np.matrix([[ 1, 4, 9, 121, 144, 169],
                [ 16, 25, 36, 196, 225, 256],
                [ 49, 64, 81, 289, 324, 361]])
print('Mean:',np.mean(mat))

Mean: 131.66666666666666


In [ ]:
a = np.arange(0,20)
a[2:7]

array([2, 3, 4, 5, 6])

In [ ]:
[a<0]

[array([False,  True,  True, False, False])]

In [ ]:
np.cos([0, 1, 2, 3, 4])

array([ 1.        ,  0.54030231, -0.41614684, -0.9899925 , -0.65364362])

In [ ]:
list_1 = [1,2,3]
list_2 = [4,5,6]
new_list = list_1 + list_2
array_1 = np.array([1,2,3])
array_2 = np.array([4,5,6])
new_array = array_1 + array_2
print(new_list)
print(new_array)

[1, 2, 3, 4, 5, 6]
[5 7 9]


In [ ]:
mat = np.array([[1, 2, 1], [4, 5, 9],[1, 8, 9]])
mat

array([[1, 2, 1],
       [4, 5, 9],
       [1, 8, 9]])

In [ ]:


n1 = np.matrix([[121, 144, 169], [196, 225, 256], [289, 324, 361]])
n2 = np.matrix([[1, 4, 9], [1, 5, 4], [9, 4, 8]])

dot_product = n1 @ n2

print(dot_product)

[[1786 1880 3017]
 [2725 2933 4712]
 [3862 4220 6785]]


In [ ]:
array_1 = np.array([1, 2, 3, 5, 8])
array_2 = np.array([0, 3, 4, 2, 1])
new_array = array_1+ array_2
final_array = new_array*array_1
print(final_array[2])

21


In [ ]:
np.ones([2, 2])

array([[1., 1.],
       [1., 1.]])

In [ ]:
mat[1] = 5
print(mat)

[[1 2 1]
 [5 5 5]
 [1 8 9]]


In [3]:
avg_rating_greater_than_4 = books[books['average_rating'] > 4]
num_books_with_avg_gt_4 = avg_rating_greater_than_4.shape[0]

NameError: name 'books' is not defined